src/wallet/rpc/encrypt.cpp

```diff
@@ -90,6 +90,7 @@ RPCHelpMan walletpassphrase()
     std::weak_ptr<CWallet> weak_wallet = wallet;
     pwallet->chain().rpcRunLater(strprintf("lockwallet(%s)", pwallet->GetName()), [weak_wallet, relock_time] {
         if (auto shared_wallet = weak_wallet.lock()) {
+            LOCK(shared_wallet->m_relock_mutex);
```

Comment:
In b8952bd7d3b3f745109506223f829c5dd91e638f "wallet: ensure that the passphrase is not deleted from memory when being used to rescan"

I think this is the wrong place to grab this mutex. I think it would be better to put it `CWallet::Lock` (in addition to the `LOCK(cs_wallet)` already there) as that is the underlying locking function that is called in a variety of places and so would offer stronger protection.

Reply:
I have some doubts about the proper location for grabbing the mutex, but I am not an expert in wallet code, some take my ideas with care. @achow101 's suggestion seems like a nice "natural" way to lock mutexes, and initially I agreed with it. But since this pr creates a new mutex, I see no conflict on grabbing it elsewhere, like here. Also I'm a little bit concerned that other calls to `CWallet::Lock` might unnecessarily lock this unrelated mutex, blocking execution.

Reply:
The new mutex is now grabbed in `CWallet::Lock` and also in locations where `cs_wallet` would have been grabbed first in order to prevent a deadlock.